### PR TITLE
feat: adding support for url-based credential files

### PR DESCRIPTION
**Pull request by ScruffyProdigy**

### CODE REVIEW

File: tests/test_identity_pool.py [hunk 1]
```diff
@@ -551,3 +561,36 @@ def test_refresh_with_retrieve_subject_token_error(self):
                 SUBJECT_TOKEN_JSON_FILE, "not_found"
             )
         )
+
+    @mock.patch.object(urllib.request, "urlopen", return_value=FakeResponse(
```
Comment (bojeil-google) [inline tests/test_identity_pool.py:565]: We need more tests. We will need to cover the following test cases (some you already covered):

- `test_from_info_full_options` but with url sourced `credential_source`
- `test_from_file_full_options` but with url sourced `credential_source`
- `test_retrieve_subject_token` but from url with raw text format
- `test_retrieve_subject_token` but from url with json format
- `test_retrieve_subject_token` but from url with invalid field name
- `test_retrieve_subject_token` but from url with invalid json response
- `test_retrieve_subject_token` but with url error (non-200 response)
- `test_refresh` with url-sourced cred without service account impersonation
- `test_refresh` with url-sourced cred with service account impersonation
- `test_refresh` with url-sourced cred error such as non-200 response.

**Review by bojeil-google — CHANGES_REQUESTED** [review]
Thanks for making the changes.

Summary of feedback:
- `headers` need to be honored.
- `urllib` not needed. `request` can be used instead and is provided in the `retrieve_subject_token()`
- More test coverage.

**Review by bojeil-google — CHANGES_REQUESTED** [review]
Forgot to mention, the file description and public function comments need to be updated to cover url-source credentials.

File: tests/test_identity_pool.py [hunk 2]
```diff
@@ -551,3 +564,177 @@ def test_refresh_with_retrieve_subject_token_error(self):
...
+    def test_retrieve_subject_token_from_url_json(self):
+        credential_source = {
+            "url": "http://fakeurl.com",
+            "format": {"type": "json", "subject_token_field_name": "access_token"},
+        }
+        credentials = self.make_credentials(credential_source=credential_source)
+        subject_token = credentials.retrieve_subject_token(
+            self.make_mock_request(token_data=JSON_FILE_CONTENT))
+
+        assert subject_token == JSON_FILE_SUBJECT_TOKEN
```
Comment (bojeil-google) [inline tests/test_identity_pool.py:587]: Can you also assert the request was sent with the expected parameters? eg. the expected URL, GET method and headers and no body? We don't have any coverage for this at the moment.

File: tests/test_identity_pool.py [hunk 3]
```diff
@@ -551,3 +564,177 @@ def test_refresh_with_retrieve_subject_token_error(self):
...
+
+    def test_retrieve_subject_token_from_url_not_found(self):
+        credential_source = {
+            "url": "http://fakeurl.com",
+        }
+        credentials = self.make_credentials(credential_source=credential_source)
+        with pytest.raises(exceptions.RefreshError) as excinfo:
+            credentials.retrieve_subject_token(
+                self.make_mock_request(
+                    token_status=500,
```
Comment (bojeil-google) [inline tests/test_identity_pool.py:597]: nit: since this is a not found error, maybe we can use 404?

File: tests/test_identity_pool.py [hunk 4]
```diff
@@ -551,3 +564,177 @@ def test_refresh_with_retrieve_subject_token_error(self):
...
+            self.make_mock_request(token_data=JSON_FILE_CONTENT))
+
+        assert subject_token == JSON_FILE_SUBJECT_TOKEN
+
+    def test_retrieve_subject_token_from_url_not_found(self):
+        credential_source = {
+            "url": "http://fakeurl.com",
+        }
+        credentials = self.make_credentials(credential_source=credential_source)
+        with pytest.raises(exceptions.RefreshError) as excinfo:
```
Comment (bojeil-google) [inline tests/test_identity_pool.py:594]: Can you assert the error message too?

**Review by bojeil-google — CHANGES_REQUESTED** [review]
Thanks for making the changes! This is looking really good. Just a few nits and minor change requests. Should be easy to resolve.

File: tests/test_identity_pool.py [hunk 5]
```diff
@@ -104,6 +105,15 @@ def make_mock_request(
 
         return request
 
+    @classmethod
+    def assert_credential_request_kwargs(
+        cls, request_kwargs, url=CREDENTIAL_URL
+    ):
+        assert request_kwargs["url"] == url
+        assert request_kwargs["method"] == "GET"
+        assert request_kwargs["headers"] is None
```
Comment (bojeil-google) [inline tests/test_identity_pool.py:114]: We still need a test with non-none headers. eg:
```
  "credential_source": {
    "url": "http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=...",
    "headers": {
      "Metadata": "True"
    },
    "format": {
      "type": "json",
      "subject_token_field_name": "access_token"
    }
  }

```

**Review by bojeil-google — APPROVED** [review]

File: google/auth/identity_pool.py [hunk 6]
```diff
@@ -91,8 +90,12 @@ def __init__(
             quota_project_id=quota_project_id,
             scopes=scopes,
         )
-        if isinstance(credential_source, dict):
+        if not isinstance(credential_source, dict):
```
Comment (busunkim96) [inline google/auth/identity_pool.py:93]: The docstring says a `Mapping` is accepted, so please either check for `collections.Mapping` or update the docstring.

File: google/auth/identity_pool.py [hunk 7]
```diff
@@ -117,28 +120,65 @@ def __init__(
                     )
             else:
                 self._credential_source_field_name = None
-        else:
-            self._credential_source_file = None
-        if not self._credential_source_file:
-            raise ValueError("Missing credential_source file")
+
```
Comment (busunkim96) [inline google/auth/identity_pool.py:123]: Could you also provide an example `credentials_source` in the function docstring so it's easy for folks to figure out the expected format?

File: google/auth/identity_pool.py [hunk 8]
```diff
@@ -91,8 +90,12 @@ def __init__(
...
         )
-        if isinstance(credential_source, dict):
+        if not isinstance(credential_source, dict):
+            self._credential_source_file = None
+            self._credential_source_url = None
+        else:
             self._credential_source_file = credential_source.get("file")
+            self._credential_source_url = credential_source.get("url")
             self._credential_source_headers = credential_source.get("headers")
             credential_source_format = credential_source.get("format") or {}
```
Comment (busunkim96) [inline google/auth/identity_pool.py:100]:
```suggestion
            credential_source_format = credential_source.get("format", {})
```

File: google/auth/identity_pool.py [hunk 9]
```diff
@@ -117,28 +120,65 @@ def __init__(
...
                 self._credential_source_field_name = None
-        else:
-            self._credential_source_file = None
-        if not self._credential_source_file:
-            raise ValueError("Missing credential_source file")
+
+        if self._credential_source_file and self._credential_source_url:
+            raise ValueError("Ambiguous credential_source")
+        if not self._credential_source_file and not self._credential_source_url:
+            raise ValueError("Missing credential_source")
```
Comment (busunkim96) [inline google/auth/identity_pool.py:127]: Please make this more explicit about what field(s) are causing the error

Examples:

- `Ambiguous credentials_source. 'file' is mutually exclusive with 'url'.`

- `Missing credentials_source. A 'file' or 'url' must be provided.`

**Review by busunkim96 — COMMENTED** [review]
Mostly LGTM, just left a few docs-related asks.

File: google/auth/identity_pool.py (hunk 6 above)
Comment (ScruffyProdigy) [reply google/auth/identity_pool.py]: Done.  I used collections.abc.Mapping, though, as I was getting warnings for using the regular collections library

File: google/auth/identity_pool.py (hunk 7 above)
Comment (ScruffyProdigy) [reply google/auth/identity_pool.py:137]: Done.  I added a couple of different examples

File: google/auth/identity_pool.py (hunk 8 above)
Comment (ScruffyProdigy) [reply google/auth/identity_pool.py]: Done

File: google/auth/identity_pool.py (hunk 9 above)
Comment (ScruffyProdigy) [reply google/auth/identity_pool.py]: Done

**Review by busunkim96 — APPROVED** [review]

**Comment by busunkim96** [comment]
@ScruffyProdigy If you [join the `googleapis` org](https://opensource.corp.google.com/github/orgs) CI will run for you automatically.
